### PR TITLE
Dataset exploration

### PR DESCRIPTION
**Pull request by corochann**
This PR adds jupyter notebook for dataset explanation.

You may refer here as well.

https://github.com/corochann/chainer-chemistry/blob/dataset_exploration/examples/qm9/qm9_dataset_exploration.ipynb
https://github.com/corochann/chainer-chemistry/blob/dataset_exploration/examples/tox21/tox21_dataset_exploration.ipynb

### CODE REVIEW

**Comment by corochann** [comment]
Please merge #79 first.

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 1]
```diff
@@ -0,0 +1,558 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## QM9 Dataset exploration"
```
Comment (n-yoshikawa) [inline examples/qm9/qm9_dataset_exploration.ipynb]: what is this?

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This notebook is executed on 16, Jan, 2018. \n",
+    "\n",
+    "The tox21 dataset behavior or `chainer_chemistry` library behavior might change in the future."
```
Comment (n-yoshikawa) [inline examples/qm9/qm9_dataset_exploration.ipynb]: How about "The behavior of tox21 dataset in `chainer_chemistry` might change in the future."?

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+    "interact(show_dataset, index=(0, len(dataset) - 1, 1))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Appendix: how to save the molecule figure?\n",
+    "\n",
+    "`rdkit` provides `Draw.MolToFile` method to visualize mol instance and save it to png format."
```
Comment (n-yoshikawa) [inline examples/qm9/qm9_dataset_exploration.ipynb]: You can save molecule images in SVG with RDKit. See http://cheminformist.itmol.com/TEST/wp-content/uploads/2015/07/rdkit_moldraw2d_2.html

File: examples/tox21/tox21_dataset_exploration.ipynb [hunk 4]
```diff
@@ -0,0 +1,650 @@
...
+    "logging.basicConfig(level=logging.INFO)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Tox 21 dataset consists of train/validation/test data and they can be downloaded automatically with chainer chemistry. \n",
+    "The format of tox21 dataset is \"sdf\" file.\n",
+    "You may check the file path of downloaded sdf file with `get_tox21_filepath` method. "
```
Comment (n-yoshikawa) [inline examples/tox21/tox21_dataset_exploration.ipynb]: You "can"?

File: examples/tox21/tox21_dataset_exploration.ipynb [hunk 5]
```diff
@@ -0,0 +1,650 @@
...
+    "Dataset extraction depends on the preprocessing method, which is determined by `preprocessor`.\n",
+    "\n",
+    "Here, let's look an example of using `NFPPreprocessor` preprocessor for tox21 dataset exraction.\n",
+    "\n",
+    "Procedure is as follows,\n",
+    "\n",
+    "1. Instantiate `preprocessor` (here `NFPPreprocessor` is used).\n",
+    "2. call `get_tox21` method with `preprocessor`.\n",
+    "\n",
+    "Note that `retrain_smiles` option can be used to get smiles information together with the dataset itself."
```
Comment (n-yoshikawa) [inline examples/tox21/tox21_dataset_exploration.ipynb]: "smiles" -> "SMILES"

File: examples/tox21/tox21_dataset_exploration.ipynb [hunk 6]
```diff
@@ -0,0 +1,650 @@
...
+      "INFO:chainer_chemistry.dataset.parsers.sdf_file_parser:Preprocess finished. FAIL 0, SUCCESS 645, TOTAL 645\n"
+     ]
+    }
+   ],
+   "source": [
+    "from chainer_chemistry.dataset.preprocessors.nfp_preprocessor import \\\n",
+    "    NFPPreprocessor\n",
+    "\n",
+    "preprocessor = NFPPreprocessor()\n",
+    "train, val, test, train_smiles, val_smiles, test_smiles = datasets.get_tox21(preprocessor, labels=None, retain_smiles=True)"
```
Comment (n-yoshikawa) [inline examples/tox21/tox21_dataset_exploration.ipynb]: This causes a lot of "RDKit ERROR".

File: examples/tox21/tox21_dataset_exploration.ipynb [hunk 7]
```diff
@@ -0,0 +1,650 @@
...
+    " 2. adjacency matrix feature: representing adjacency matrix of given molecule.\n",
+    " 3. label feature: representing toxity (label) of given molecule.\n",
+    "    Here, 0 indicates negative (no toxity), 1 indicates positive (toxic) and -1 indicates data is not available, respectively."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's look an example of 7777-th train dataset"
```
Comment (n-yoshikawa) [inline examples/tox21/tox21_dataset_exploration.ipynb]: The result was different in my environment.

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "dataset_filepath = datasets.get_qm9_filepath()\n",
+    "\n",
+    "# print('dataset_filepath = ', dataset_filepath)"
```
[the remark this answers is not in the document]
Comment (corochann) [reply examples/qm9/qm9_dataset_exploration.ipynb]: noted, I'll remove all space of sentence with "="

File: examples/qm9/qm9_dataset_exploration.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (corochann) [reply examples/qm9/qm9_dataset_exploration.ipynb]: ok

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The dataset contains several chemical/physical properties, the labels of QM 9 dataset can be checked by `get_qm9_label_names` method."
```
[the remark this answers is not in the document]
Comment (corochann) [reply examples/qm9/qm9_dataset_exploration.ipynb]: ok

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+   "source": [
+    "label_names = datasets.get_qm9_label_names()\n",
+    "print('QM9 label_names = ', label_names)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "More detail information is described in `readme.txt` of QM 9 dataset, which can be downloaded from \n",
```
[the remark this answers is not in the document]
Comment (corochann) [reply examples/qm9/qm9_dataset_exploration.ipynb]: noted

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Preprocessing dataset\n",
+    "\n",
+    "Dataset extraction depends on the preprocessing method, which is determined by `preprocessor`.\n",
+    "\n",
+    "Here, let's look an example of using `GGNNPreprocessor` preprocessor for QM9 dataset exraction.\n",
```
[the remark this answers is not in the document]
Comment (corochann) [reply examples/qm9/qm9_dataset_exploration.ipynb]: ok

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 12]
```diff
@@ -0,0 +1,558 @@
...
+    "\n",
+    "Procedure is as follows,\n",
+    "\n",
+    "1. Instantiate `preprocessor` (here `GGNNPreprocessor` is used).\n",
+    "2. call `get_qm9` method with `preprocessor`.\n",
+    "\n",
+    "Note that `retrain_smiles` option can be used to get smiles information together with the dataset itself."
```
[the remark this answers is not in the document]
Comment (corochann) [reply examples/qm9/qm9_dataset_exploration.ipynb]: ok

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 13]
```diff
@@ -0,0 +1,558 @@
...
+    "    `GGNNPreprocessor` extracts adjacency matrix of each bonding type.\n",
+    " 3. label feature: representing chemical properties (label) of given molecule.\n",
+    "    Please refer above description for details."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's look an example of 7777-th dataset"
```
[the remark this answers is not in the document]
Comment (corochann) [reply examples/qm9/qm9_dataset_exploration.ipynb]: Added link

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 14]
```diff
@@ -0,0 +1,558 @@
...
+   "metadata": {},
+   "source": [
+    "The dataset is a class of `NumpyTupleDataset`, where i-th dataset features can be accessed by `dataset[i]`.\n",
+    "\n",
+    "When `GGNNPreprocessor` is used, each dataset consists of following features\n",
+    " 1. atom feature: representing atomic number of given molecule. \n",
+    " 2. adjacency matrix feature: representing adjacency matrix of given molecule.\n",
+    "    `GGNNPreprocessor` extracts adjacency matrix of each bonding type.\n",
+    " 3. label feature: representing chemical properties (label) of given molecule.\n",
+    "    Please refer above description for details."
```
[the remark this answers is not in the document]
Comment (corochann) [reply examples/qm9/qm9_dataset_exploration.ipynb]: Added link

File: examples/qm9/qm9_dataset_exploration.ipynb [hunk 15]
```diff
@@ -0,0 +1,558 @@
...
+    "print('adjacency matrix for AROMATIC bond type\\n', adj[3])\n",
+    "\n",
+    "print('labels', labels)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Visualizing the molecule\n",
```
[the remark this answers is not in the document]
Comment (corochann) [reply examples/qm9/qm9_dataset_exploration.ipynb]: thanks

File: examples/tox21/tox21_dataset_exploration.ipynb [hunk 16]
```diff
@@ -0,0 +1,650 @@
...
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This notebook is executed on 16, Jan, 2018. \n",
+    "\n",
+    "The tox21 dataset behavior or `chainer_chemistry` library behavior might change in the future."
```
[the remark this answers is not in the document]
Comment (corochann) [reply examples/tox21/tox21_dataset_exploration.ipynb]: I meant that tox21 comes from outside source, so behavior might change due to outside situation not our fault.

Well but I will adopt above sentence.

File: examples/tox21/tox21_dataset_exploration.ipynb (hunk 4 above)
Comment (corochann) [reply examples/tox21/tox21_dataset_exploration.ipynb]: ok

File: examples/tox21/tox21_dataset_exploration.ipynb (hunk 5 above)
Comment (corochann) [reply examples/tox21/tox21_dataset_exploration.ipynb]: ok

File: examples/tox21/tox21_dataset_exploration.ipynb (hunk 6 above)
Comment (corochann) [reply examples/tox21/tox21_dataset_exploration.ipynb]: Added log suppression code

File: examples/tox21/tox21_dataset_exploration.ipynb (hunk 7 above)
Comment (corochann) [reply examples/tox21/tox21_dataset_exploration.ipynb]: let's discuss about it.

**Comment by corochann** [comment]
Updated based on comments.

File: examples/tox21/tox21_dataset_exploration.ipynb [hunk 17]
```diff
@@ -174,11 +174,11 @@
      "name": "stderr",
      "output_type": "stream",
      "text": [
-      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 11764/11764 [00:23<00:00, 501.06it/s]\n",
+      "100%|███████████████████████████████████████████████████████████████████████████| 11764/11764 [00:29<00:00, 401.74it/s]\n",
       "INFO:chainer_chemistry.dataset.parsers.sdf_file_parser:Preprocess finished. FAIL 0, SUCCESS 11757, TOTAL 11757\n",
```
Comment (n-yoshikawa) [inline examples/tox21/tox21_dataset_exploration.ipynb]: RDKit version also affects this result. In newer version, `SUCCESS 11757` becomes `SUCCESS 11758`. Therefore, the note about RDKit version should come before here.

File: examples/tox21/tox21_dataset_exploration.ipynb [hunk 18]
```diff
@@ -0,0 +1,769 @@
...
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "100%|███████████████████████████████████████████████████████████████████████████| 11764/11764 [00:22<00:00, 531.76it/s]\n",
+      "INFO:chainer_chemistry.dataset.parsers.sdf_file_parser:Preprocess finished. FAIL 0, SUCCESS 11757, TOTAL 11757\n",
```
Comment (n-yoshikawa) [inline examples/tox21/tox21_dataset_exploration.ipynb]: RDKit version also affects this result. In newer version, `SUCCESS 11757` becomes `SUCCESS 11758`. Therefore, the note about RDKit version should come before here.

File: examples/tox21/tox21_dataset_exploration.ipynb [hunk 19]
```diff
@@ -0,0 +1,769 @@
...
+     "output_type": "stream",
+     "text": [
+      "index=5878, SMILES=CN(C)CCn1nnnc1SCC1=C(C(=O)O)N2C(=O)C(NC(=O)Cc3csc(N)n3)C2SC1.Cl\n",
+      "atom [ 6  7  6  6  6  7  7  7  7  6 16  6  6  6  6  8  8  7  6  8  6  7  6  8\n",
+      "  6  6  6 16  6  7  7  6 16  6 17]\n",
+      "labels [ 0  0  0  0  0  0  0 -1  0 -1  0  0]\n"
+     ]
+    },
+    {
+     "data": {
```
Comment (n-yoshikawa) [inline examples/tox21/tox21_dataset_exploration.ipynb]: SVG images cannot be displayed on GitHub. Maybe you should note about it.

File: examples/tox21/tox21_dataset_exploration.ipynb [hunk 20]
```diff
@@ -0,0 +1,769 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Check extracted dataset\n",
+    "\n",
+    "First, let's check number of data for train/validation/test dataset.\n",
+    "\n",
+    "[Note] Preprocessing result depends on Rdkit version. \n",
```
Comment (n-yoshikawa) [inline examples/tox21/tox21_dataset_exploration.ipynb]: Rdkit should be RDKit.

**Pull request merged**